tensorflow/python/training/loss_scale_optimizer.py

```diff
@@ -0,0 +1,235 @@
+        `GATE_NONE`, `GATE_OP`, or `GATE_GRAPH`.
+      aggregation_method: Specifies the method used to combine gradient terms.
+        Valid values are defined in the class `AggregationMethod`.
+      colocate_gradients_with_ops: If True, try colocating gradients with
+        the corresponding op.
+      grad_loss: Optional. A `Tensor` holding the gradient computed for `loss`.
+
+    Returns:
+      A list of (gradient, variable) pairs. Variable is always present, but
+      gradient can be `None`.
+    """
+    loss = self._scale_loss(loss)
+    grads_and_vars = self._optimizer.compute_gradients(
+        loss=loss, var_list=var_list, gate_gradients=gate_gradients,
+        aggregation_method=aggregation_method,
+        colocate_gradients_with_ops=colocate_gradients_with_ops,
+        grad_loss=grad_loss)
+
+    grads = [g for g, _ in grads_and_vars]
+    variables = [v for _, v in grads_and_vars]
+    scaled_grads = self._scale_grads(grads)
+    return list(zip(scaled_grads, variables))
+
+  def _scale_loss(self, loss):
+    # The loss is callable for `_compute_gradients`, but not `get_gradients`.
+    loss_scale = self._loss_scale()
+    if callable(loss):
+      return lambda: loss() * loss_scale
+    return loss * loss_scale
+
+  def _scale_grads(self, grads):
+    loss_scale = self._loss_scale()
+    loss_scale_reciprical = 1 / loss_scale
+    return [None if g is None else self._indexed_slices(
+        g, loss_scale_reciprical) for g in grads]
+
+  def _indexed_slices(self, grad, loss_scale_reciprical):
+    if isinstance(grad, ops.IndexedSlices):
+      grad_vals = grad.values * loss_scale_reciprical
+      return ops.IndexedSlices(grad_vals, grad.indices, grad.dense_shape)
+    return grad * loss_scale_reciprical
+
+  def apply_gradients(self, grads_and_vars, global_step=None, name=None):
+    """Apply gradients to variables.
+
+    This is the second part of `minimize()`. It returns an `Operation` that
+    conditionally applies gradients if all gradient values are finite.
+    Otherwise no update is performed (nor is `global_step` incremented).
+
+    Args:
+      grads_and_vars: List of (gradient, variable) pairs as returned by
+        `compute_gradients()`.
+      global_step: Optional `Variable` to increment by one after the
+        variables have been updated.
+      name: Optional name for the returned operation.  Default to the name
+        passed to the `Optimizer` constructor.
+
+    Returns:
+      An `Operation` that conditionally applies the specified gradients. If
+      `global_step` was not None, that operation also increments `global_step`.
+
+    Raises:
+      RuntimeError: If you should use `_distributed_apply()` instead.
+    """
+    if distribution_strategy_context.in_cross_replica_context():
+      raise ValueError('apply_gradients() must be called in a replica context.')
+
+    if not self._doing_dynamic_loss_scaling():
+      return self._optimizer.apply_gradients(grads_and_vars, global_step, name)
```

Comment:
Why have a separate case here? Won't the codepath below also handle fixed loss scaling?

Reply:
The intent was to minimize overheads, but the below code should handle fixed as well